internal/prefixcollector/collector.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package prefixcollector contains excluded prefix collector, prefix sources and functions working with them
+package prefixcollector
+
+import (
+	"cmd-exclude-prefixes-k8s/internal/utils"
+	"context"
+	"io/ioutil"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/excludedprefixes"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/prefixpool"
+)
+
+// ExcludePrefixSource is source of excluded prefixes
+type ExcludePrefixSource interface {
+	Prefixes() []string
+}
+
+// ExcludePrefixCollector is service, collecting excluded prefixes from list of ExcludePrefixSource
+// and environment variable "EXCLUDED_PREFIXES"
+// and writing result to outputFilePath in yaml format
+type ExcludePrefixCollector struct {
+	notifyChan          chan struct{}
+	baseExcludePrefixes []string
+	outputFilePath      string
+	sources             []ExcludePrefixSource
+}
+
+// ExcludePrefixCollectorOption is functional option for ExcludePrefixCollector
+type ExcludePrefixCollectorOption func(*ExcludePrefixCollector)
+
+const (
+	// DefaultConfigMapNamespace is default namespace of kubernetes ConfigMap
+	DefaultConfigMapNamespace = "default"
+	defaultConfigMapName      = "nsm-config-volume"
+	outputFilePermissions     = 0600
+)
+
+// WithFilePath returns ExcludePrefixCollectorOption, that set filePath as collector's outputFilePath
+func WithFilePath(filePath string) ExcludePrefixCollectorOption {
+	return func(collector *ExcludePrefixCollector) {
+		collector.outputFilePath = filePath
+	}
+}
+
+// WithSources returns ExcludePrefixCollectorOption, that set sources as collector's prefix sources
+func WithSources(sources ...ExcludePrefixSource) ExcludePrefixCollectorOption {
+	return func(collector *ExcludePrefixCollector) {
+		collector.sources = sources
+	}
+}
+
+// NewExcludePrefixCollector creates ExcludePrefixCollector
+// and applies every ExcludePrefixCollectorOption to it
+func NewExcludePrefixCollector(ctx context.Context, prefixesFromEnv []string,
```

Comment:
I do not think With options are required for two parameters, it should Please also create a source for env prefixes aka 'baseExcludePrefixes' so 

```suggestion
func NewExcludePrefixCollector(ctx context.Context, outputFile string, sources []ExcludePrefixSource, notify *sync.Cond)
```

Please use sync.Cond to notify about event.

Reply:
Options were removed. If we use env prefixes source with sync.Cond, we need to notify listeners in loop. Is it ok?

Reply:
Fixed without loop.